feat: approve/reject lector and center

siniphia opened the pull request:
## Related issue
resolves #23 

## Description
- 4 new APIs approving and rejecting lector and center are implemented.

## Changes detail

- 
- 

### Checklist

- [X] Test case
- [X] End of work

siniphia committed:
feat: jwt & user sign

siniphia committed:
feat: JWT filter dependency injection

siniphia committed:
refac: merge jwt functions into single one

- all 'jwtsample.py' and related files should be deprecated in the future
- sample endpoints are given
1. /sign-up : simply generate dummy user
2. /sign-in : login and get initial jwt tokens
3. /hello : auth with your tokens

siniphia committed:
chore: add passlib for pytest

siniphia committed:
refac: delete unnecessary codes & refator jwt logic

siniphia committed:
refac: add error code & modify create token parameter

siniphia committed:
test: add sample user repository test

siniphia committed:
chore: add python-jose for jwt filter

siniphia committed:
fix: add create-drop mode to local and test db

siniphia committed:
refac: jwt filter to return user profile & to add tokens to response header

siniphia committed:
chore: set coverage range

siniphia committed:
Update pyproject.toml

siniphia committed:
chore: codecov setting

siniphia committed:
refac: separate  jwt config

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
feat: nickname duplication API

siniphia committed:
feat: sign-up lector and center admin

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
fix: do not store proof files

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
refac: separate approved file & isolate fixtures

siniphia committed:
refac: sign-up refacoring by reviews

siniphia committed:
refac: optimize business logic & change field validators

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
feat: approve/reject lector and center APIs

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
Update center.py

pingoo33 commented on claon_admin/common/error/exception.py:
```diff
@@ -6,6 +6,8 @@ class ErrorCode(Enum):
     USER_ALREADY_SIGNED_UP = 40001
     DUPLICATED_NICKNAME = 40002
+    NONE_ADMIN_ACCOUNT = 40003
+    ENTITY_NOT_FOUND = 40004
```
NONE_ADMIN_ACCONT는 권한 관련 오류이며, 401 Unauthorized error와 관련된 것으로 보이며,
ENTITY_NOT_FOUND는 리소스 관련 오류이며, 404 Not Found와 관련된 것으로 보입니다.

pingoo33 commented on claon_admin/model/center.py:
```diff
@@ -93,6 +93,8 @@ class CenterNameResponseDto(BaseModel):
 
 class CenterResponseDto(BaseModel):
+    center_id: str
+    profile: UserProfileResponseDto
```
center response에 user profile response dto를 추가하신 이유가 있을까요?

pingoo33 replied:
만약 admin을 위해 만드신 것이라면, admin을 위한 별도의 response dto를 만드는 것이 좋을 것 같습니다.
별도의 파일을 만들어도 좋을 것 같네요!

pingoo33 commented on claon_admin/router/admin.py:
```diff
@@ -1,19 +1,24 @@
 class AdminRouter:
-    def __init__(self):
-        pass
+    def __init__(self,
```
inject 어노테이션이 필요합니다.
더불어 main.py 에서 wiring도 필요합니다.

pingoo33 commented on claon_admin/service/user.py:
```diff
@@ -132,6 +135,72 @@ async def sign_up_lector(self, session: AsyncSession, subject: RequestUser, dto:
+                "어드민 권한이 없습니다."
+            )
+
+        if not await self.lector_repository.exists_by_id(session, lector_id):
```
현재 query가 exist + find + update 이렇게 세 번 이뤄지는 것으로 보입니다.
find를 한 번하고 해당 영속 객체에 update를 하는 것으로 두 번으로 진행할 수 있을 것으로 보입니다!

pingoo33 replied:
해당 리뷰는 center approve도 적용 가능합니다.

pingoo33 commented on claon_admin/service/user.py:
```diff
@@ -132,6 +135,72 @@ async def sign_up_lector(self, session: AsyncSession, subject: RequestUser, dto:
 
         return LectorResponseDto.from_entity(lector)
 
+    async def approve_lector(self, lector_id: str, session: AsyncSession, subject: RequestUser):
```
모든 approve 및 reject 이후에는 approvedFile을 DB 및 S3에서 삭제해야 합니다.

siniphia replied:
reject 할때는 cascade option에 의해 제거되게 만들었습니다

pingoo33 commented on claon_admin/router/admin.py:
```diff
@@ -1,19 +1,24 @@
-        pass
+    def __init__(self,
+                 user_service: UserService = Depends(Provide[Container.user_service])):
+        self.user_service = user_service
```
approve 및 reject 관련 내용이 모두 user service에 있는 것이 자연스러울까요??
center 쪽은 center service에 있는 것이 자연스러울수도 있을 것 같습니다.

혹시 approve 및 reject이 나누어져 있는게 어색하다면 admin service를 새로 만드는 것도 좋을 것 같습니다.

pingoo33 commented on tests/repository/fixtures.py:
```diff
@@ -19,7 +19,7 @@
 
 
-@pytest.fixture(scope="session")
+@pytest.fixture(autouse=True)
```
제가 알기로는 이렇게 conftest에 이어 붙인 파일에서 autouse를 사용하면 모든 테스트 함수에서 해당 fixture를 사용하지 않아도 자동 호출되는 것으로 알고 있습니다.

필요한 fixture만 사용할 수 있도록 리팩토링이 필요합니다.

siniphia committed:
refac: optimize service query and some fixes

- error code category
- minimize repository function call
- delete user information from center and lector response dtos

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop

siniphia committed:
refac: repackage admin service and test code

siniphia committed:
fix: import error

siniphia committed:
refactor: s3 deletion and minor fixes

- delete approved files from s3
- change error code
- change approved file cascade option

pingoo33 commented on claon_admin/common/util/s3.py:
```diff
@@ -39,3 +40,10 @@ async def upload_file(file: UploadFile, domain: str, purpose: str):
+
+async def delete_file(url: str):
+    try:
+        client_s3.delete_object(Bucket=BUCKET, Key=os.path.basename(url))
```
객체 키는 `https://" + BUCKET + ".s3." + REGION_NAME + ".amazonaws.com`를 제외한 파일 전체 경로가 되어야 합니다. 확인 부탁드려요.

pingoo33 commented on claon_admin/model/admin.py:
```diff
@@ -1,9 +1,12 @@
-from datetime import date
+from datetime import date, datetime
```
datetime을 해당 파일에서 사용하지 않는 것 같습니다.

pingoo33 commented on claon_admin/model/center.py:
```diff
@@ -251,40 +251,42 @@ class CenterResponseDto(BaseModel):
+            tel=entity.tel,
+            web_url=entity.web_url,
+            instagram_name=entity.instagram_name,
+            youtube_code=str(entity.youtube_url).split("/")[-1],
```
저희 youtube_url이 nullable한데 null-safe 하지 않은 코드인 것 같습니다.

pingoo33 commented on tests/repository/test_user.py:
```diff
@@ -193,6 +337,30 @@ async def test_save_all_lector_approved_files(
 @pytest.mark.asyncio
+async def test_find_all_lector_approved_files_by_lector_id(
+        session: AsyncSession,
+        lector_fixture,
```
type hint가 필요합니다.

siniphia replied:
타입힌트 없는 테스트 코드에 모두 추가했습니다

siniphia committed:
fix: null-safe dto fields, s3 key name

siniphia committed:
Merge branch 'develop' of https://github.com/Co-Laon/claon-admin-server into develop
